more_itertools/more.py

```diff
@@ -1654,35 +1654,145 @@ def numeric_range(*args):
+        else:
+            if self._start >= elem > self._stop:
+                return (self._start - elem) % (-self._step) == self._zero
+
+        return False
+
+    def __eq__(self, other):
+        return (self._start == other._start
+                and self._step == other._step
+                and self._get_by_index(-1) == other._get_by_index(-1))
+
+    def __getitem__(self, key):
+        if isinstance(key, int):
+            return self._get_by_index(key)
+        elif isinstance(key, slice):
+            start = self._start if key.start is None \
+                else self._get_by_index(key.start)
+            stop = self._stop if key.stop is None \
+                else self._get_by_index(key.stop)
+            step = self._step if key.step is None else key.step * self._step
+            return numeric_range(start, stop, step)
+        else:
+            raise TypeError(
+                'numeric range indices must be '
+                'integers or slices, not {}'.format(type(key).__name__))
+
+    def __hash__(self):
+        if self:
+            return hash((self._start, self._get_by_index(-1), self._step))
+        else:
+            return self._EMPTY_HASH
+
+    def __iter__(self):
+        values = (self._start + (self._step * n) for n in count())
+        if self._growing:
+            return takewhile(partial(gt, self._stop), values)
+        else:
+            return takewhile(partial(lt, self._stop), values)
+
+    def __len__(self):
+        if self._growing:
+            start = self._start
+            stop = self._stop
+            step = self._step
+        else:
+            start = self._stop
+            stop = self._start
+            step = -self._step
+
+        distance = stop - start
+        if distance <= 0:
+            return 0
+
+        # distance > 0 and step > 0: regular euclidean division
+        q, r = divmod(distance, step)
+        return int(q) + int(r != self._zero)
+
+    def __reduce__(self):
+        return numeric_range, (self._start, self._stop, self._step)
+
+    def __repr__(self):
+        if self._step == 1:
+            return "numeric_range({}, {})".format(repr(self._start),
+                                                  repr(self._stop))
+        else:
+            return "numeric_range({}, {}, {})".format(repr(self._start),
+                                                      repr(self._stop),
+                                                      repr(self._step))
+
+    def __reversed__(self):
+        return iter(numeric_range(self._get_by_index(-1),
+                                  self._start - self._step, -self._step))
+
+    def count(self, value):
+        return int(value in self)
+
+    def index(self, value):
+        if self._growing:
+            if self._start <= value < self._stop:
+                q, r = divmod(value - self._start, self._step)
+                if r == self._zero:
+                    return int(q)
+        else:
```

Comment:
I don't think this branch is tested - could you add something that verifies this?

Reply:
Done.